### ISSUE TITLE
SAC + LSTM

### ISSUE THREAD

**Issue by jaried** [aai-institute/tianshou#717]
-  I have visited the [source website](https://github.com/thu-ml/tianshou/)
-  I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
-  I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, gym, torch, numpy, sys
  print(tianshou.__version__, gym.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
0.4.9 0.24.1 1.11.0+cu113 1.22.4 3.9.0 (default, Nov 15 2020, 08:30:55) [MSC v.1916 64 bit (AMD64)] win32
  ```

I use `sac + LSTM` to play `Pendulum-v1`, but the target value has not been reached, and the reward has not increased. Is there any problem with my code?
Purple is `sac+LSTM`, green is normal `sac`.

![1659867195412](https://user-images.githubusercontent.com/26672425/183286052-b80f3eab-f3d1-4c07-a040-0bc8560f0d3c.png)
![1659867222556](https://user-images.githubusercontent.com/26672425/183286067-7b8a65b9-2957-4286-b1a3-19671d2b9444.png)
![1659867240236](https://user-images.githubusercontent.com/26672425/183286071-b65f5e85-7c90-4021-9538-eae80ae4cbae.png)

My code is as follows:

```python
import argparse
import os

import numpy as np
import pytest
import gym
import torch
from torch.utils.tensorboard import SummaryWriter

from tianshou.data import Collector, VectorReplayBuffer
from tianshou.policy import ImitationPolicy, SACPolicy
from tianshou.trainer import offpolicy_trainer
from tianshou.utils import TensorboardLogger
from tianshou.utils.net.common import Net
from tianshou.utils.net.continuous import Actor, RecurrentActorProb, RecurrentCritic, ActorProb, Critic
from tianshou.env import DummyVectorEnv, SubprocVectorEnv
from datetime import datetime

try:
    import envpool
except ImportError:
    envpool = None


def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('--task', type=str, default='Pendulum-v1')
    parser.add_argument('--reward-threshold', type=float, default=None)
    parser.add_argument('--seed', type=int, default=1)
    parser.add_argument('--buffer-size', type=int, default=20000)
    parser.add_argument('--actor-lr', type=float, default=1e-3)
    parser.add_argument('--critic-lr', type=float, default=1e-3)
    parser.add_argument('--il-lr', type=float, default=1e-3)
    parser.add_argument('--gamma', type=float, default=0.99)
    parser.add_argument('--tau', type=float, default=0.005)
    parser.add_argument('--alpha', type=float, default=0.2)
    parser.add_argument('--auto-alpha', type=int, default=1)
    parser.add_argument('--alpha-lr', type=float, default=3e-4)
    parser.add_argument('--epoch', type=int, default=100000)
    parser.add_argument('--step-per-epoch', type=int, default=24000)
    parser.add_argument('--il-step-per-epoch', type=int, default=500)
    parser.add_argument('--step-per-collect', type=int, default=10)
    parser.add_argument('--update-per-step', type=float, default=0.1)
    parser.add_argument('--batch-size', type=int, default=128)
    parser.add_argument('--hidden-sizes', type=int, nargs='*', default=[128, 128])
    parser.add_argument(
        '--imitation-hidden-sizes', type=int, nargs='*', default=[128, 128]
    )
    parser.add_argument('--training-num', type=int, default=10)
    parser.add_argument('--test-num', type=int, default=100)
    parser.add_argument('--logdir', type=str, default='log')
    parser.add_argument('--render', type=float, default=0.)
    parser.add_argument('--rew-norm', action="store_true", default=False)
    parser.add_argument('--n-step', type=int, default=3)
    parser.add_argument(
        '--device', type=str, default='cuda' if torch.cuda.is_available() else 'cpu'
    )
    args = parser.parse_known_args()[0]
    return args


# @pytest.mark.skipif(envpool is None, reason="EnvPool doesn't support this platform")
def sac_with_il(args):
    # if you want to use python vector env, please refer to other test scripts
    start_datetime = datetime.now()
    env = gym.make(args.task)
    train_envs = DummyVectorEnv(
        [lambda: gym.make(args.task) for _ in range(args.training_num)]
    )
    # test_envs = DummyVectorEnv(
    #     [lambda: gym.make(args.task) for _ in range(args.test_num)]
    # )
    args.state_shape = env.observation_space.shape or env.observation_space.n
    args.action_shape = env.action_space.shape or env.action_space.n
    args.max_action = env.action_space.high[0]
    if args.reward_threshold is None:
        default_reward_threshold = {"Pendulum-v0": -250, "Pendulum-v1": -250}
        args.reward_threshold = default_reward_threshold.get(
            args.task, env.spec.reward_threshold
        )
    # you can also use tianshou.env.SubprocVectorEnv
    # seed
    np.random.seed(args.seed)
    torch.manual_seed(args.seed)
    # model
    layer_num = 4
    actor = RecurrentActorProb(
        layer_num=layer_num,
        state_shape=args.state_shape,
        action_shape=args.action_shape,
        max_action=args.max_action,
        device=args.device,
        unbounded=True
    ).to(args.device)
    actor_optim = torch.optim.Adam(actor.parameters(), lr=args.actor_lr)
    hidden_layer_size = 128
    critic1 = RecurrentCritic(
        layer_num=layer_num,
        state_shape=args.state_shape,
        action_shape=args.action_shape,
        hidden_layer_size=hidden_layer_size,
        device=args.device).to(args.device)
    critic1_optim = torch.optim.Adam(critic1.parameters(), lr=args.critic_lr)

    critic2 = RecurrentCritic(
        layer_num=layer_num,
        state_shape=args.state_shape,
        action_shape=args.action_shape,
        hidden_layer_size=hidden_layer_size,
        device=args.device
    ).to(args.device)
    critic2_optim = torch.optim.Adam(critic2.parameters(), lr=args.critic_lr)

    if args.auto_alpha:
        target_entropy = -np.prod(env.action_space.shape)
        log_alpha = torch.zeros(1, requires_grad=True, device=args.device)
        alpha_optim = torch.optim.Adam([log_alpha], lr=args.alpha_lr)
        args.alpha = (target_entropy, log_alpha, alpha_optim)

    policy = SACPolicy(
        actor,
        actor_optim,
        critic1,
        critic1_optim,
        critic2,
        critic2_optim,
        tau=args.tau,
        gamma=args.gamma,
        alpha=args.alpha,
        reward_normalization=args.rew_norm,
        estimation_step=args.n_step,
        action_space=env.action_space
    )
    # collector
    train_collector = Collector(
        policy,
        train_envs,
        VectorReplayBuffer(args.buffer_size, len(train_envs), stack_num=4),
        exploration_noise=True
    )
    test_collector = None
    # train_collector.collect(n_step=args.buffer_size)
    # log
    log_path = os.path.join(args.logdir, args.task, 'sac', start_datetime.strftime('%Y%m%d/%H%M%S'))
    writer = SummaryWriter(log_path)
    logger = TensorboardLogger(writer)

    def save_best_fn(policy):
        torch.save(policy.state_dict(), os.path.join(log_path, 'policy.pth'))

    def stop_fn(mean_rewards):
        return mean_rewards >= args.reward_threshold

    # trainer
    train_collector.collect(n_step=20000, random=True)
    result = offpolicy_trainer(
        policy,
        train_collector,
        test_collector,
        args.epoch,
        args.step_per_epoch,
        args.step_per_collect,
        args.test_num,
        args.batch_size,
        update_per_step=args.update_per_step,
        stop_fn=stop_fn,
        save_best_fn=save_best_fn,
        logger=logger
    )


if __name__ == '__main__':
    args1 = get_args()
    sac_with_il(args1)

```

**Comment by Trinkle23897** [comment]
Unfortunately, there's a known issue in the current codebase #486 that has not been fixed. I think that should be the root cause :(

**Comment by jaried** [comment]
Excuse me, is there any plan to fix this bug? If there is no plan, I will temporarily close the issue.

**Comment by Trinkle23897** [comment]
no plan :(

**Comment by jaried** [comment]
Okay, I understand now, thanks.

**Comment by 1900360** [comment]
Is there any workaround for this? I also had the same problem
